### DISCUSSION TITLE
JumpHost using a server

### DISCUSSION THREAD

**Discussion by aferre** [gravitational/teleport#9011] [Q&A] [answered]
Hi,

I have a setup containing a node (named vpn-strongswan) and I would like to use it as a jumphost to reach clients through site-to-site VPN. I have tried using a few different commands and have failed to access a remote machine.
   
![image](https://user-images.githubusercontent.com/570643/141960414-60809a74-79e7-41f1-b4ef-5e8e3c7bb1b9.png)

When I try using the following command I can reach the machine (SSH only)

`ssh -J IP_OF_NODE USER_OF_MACHINE@IP_OF_MACHINE`

When trying to do the same thing using teleport, I am not able to reach the machine and the errors provided by the tsh client on my machine all relates to authentication failures (which is weird because I can reach the proxy and the node using tsh ssh without issue)

`tsh ssh -J vpn-strongswan USER_OF_MACHINE@IP_OF_MACHINE`

The following error is provided by tsh

`ERROR: failed to authenticate with proxy vpn-strongswan
dial tcp: address vpn-strongswan: missing port in address`

I added the full domain name for the node with the port (3022). tsh is now using this node as a proxy for authentication. If I specify the relevant proxy for the authentication, tsh still tries to use the node as a proxy and tsh still prompts the same error.

`tsh ssh -J DOMAIN_NAME_OF_NODE:3022 --proxy=DOMAIN_NAME_OF_PROXY_SERVER admin@IP_OF_MACHINE`

Is there something I am missing ?

**Comment by webvictim** [comment, marked as answer]
It looks like this is intentional behaviour - https://github.com/gravitational/teleport/blob/master/lib/client/api.go#L2042-L2044

Basically if you run `tsh ssh -J host1 host2`, `tsh` will try to use `host1` to connect to the Teleport proxy before jumping on to `host2`, rather than what you're expecting - connecting directly to the proxy, then connecting to `host1` to make the connection to `host2`.

My advice here would be to write an SSH config which does what you want:

```console
Host teleport.example.com
    User exampleuser
    Port 3023
    IdentityFile ~/.tsh/keys/teleport.example.com/teleportuser
    CertificateFile ~/.tsh/keys/teleport.example.com/teleportuser-ssh/teleport.example.com-cert.pub

Host host1
    User exampleuser
    ProxyCommand 'ssh teleport.example.com -s proxy:%h:%p'
    Port 3022
    IdentityFile ~/.tsh/keys/teleport.example.com/teleportuser
    CertificateFile ~/.tsh/keys/teleport.example.com/teleportuser-ssh/teleport.example.com-cert.pub
```

- Replace `exampleuser` with your UNIX login on the system.
- Replace `teleportuser` with your Teleport username
- Replace `teleport.example.com` with your Teleport SSH proxy hostname

At this point you should be able to `ssh -J host1 host2` to use `host1` as a jumphost for `host2`, but have the traffic proxied via Teleport.

See https://github.com/gravitational/teleport/discussions/5530 for some more detailed instructions on using OpenSSH as a client for Teleport. You can also use `tsh config` to automatically generate an SSH config which might give you some useful pointers.

If you have issues, post the output of `ssh -vv host1 host2` and I can try to help you fix them.

**Reply by aferre** [reply]
Thank you for your help.

I have tried the setup you provide but stil failed to connect to the remote machine.

Here is the output of the  `ssh -J vpn-strongswan 172.40.1.1 -vv`  command

OpenSSH_8.2p1 Ubuntu-4ubuntu0.3, OpenSSL 1.1.1f  31 Mar 2020
debug1: Reading configuration data MY_HOME/.ssh/config
debug1: Reading configuration data MY_HOME/.ssh/teleport/ssh_config
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug2: resolve_canonicalize: hostname 172.40.1.1 is address
debug1: Setting implicit ProxyCommand from ProxyJump: ssh -vv -W '[%h]:%p' vpn-strongswan
debug1: Executing proxy command: exec ssh -vv -W '[172.40.1.1]:22' vpn-strongswan
debug1: identity file MY_HOME/.ssh/id_rsa type 0
debug1: identity file MY_HOME/.ssh/id_rsa-cert type -1
debug1: identity file MY_HOME/.ssh/id_dsa type -1
debug1: identity file MY_HOME/.ssh/id_dsa-cert type -1
debug1: identity file MY_HOME/.ssh/id_ecdsa type -1
debug1: identity file MY_HOME/.ssh/id_ecdsa-cert type -1
debug1: identity file MY_HOME/.ssh/id_ecdsa_sk type -1
debug1: identity file MY_HOME/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file MY_HOME/.ssh/id_ed25519 type -1
debug1: identity file MY_HOME/.ssh/id_ed25519-cert type -1
debug1: identity file MY_HOME/.ssh/id_ed25519_sk type -1
debug1: identity file MY_HOME/.ssh/id_ed25519_sk-cert type -1
debug1: identity file MY_HOME/.ssh/id_xmss type -1
debug1: identity file MY_HOME/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.3
OpenSSH_8.2p1 Ubuntu-4ubuntu0.3, OpenSSL 1.1.1f  31 Mar 2020
debug1: Reading configuration data MY_HOME/.ssh/config
debug1: Reading configuration data MY_HOME/.ssh/teleport/ssh_config
debug1: MY_HOME/.ssh/teleport/ssh_config line 7: Applying options for vpn-strongswan
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: Executing proxy command: exec 'ssh tele-test.dot.com -s proxy:vpn-strongswan:3022'
debug1: identity file MY_HOME/.tsh/keys/tele-test.dot.com/teleport-admin type 0
debug1: certificate file MY_HOME/.tsh/keys/tele-test.dot.com/teleport-admin-ssh/tele-test.dot.com-cert.pub type 4
debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.3
/bin/bash: line 0: exec: ssh tele-test.dot.com -s proxy:vpn-strongswan:3022: not found

This part really seems strange to me: the information you provide and the https://github.com/gravitational/teleport/discussions/5530 link you provided are stating that this should work. I could not get my head around why that would work on paper and not on my machine. 
For some reason, Teleport 6 was installed on my machine (installed it last friday)...

`teleport version` 
Teleport v6.2.18-label-debug.2 git:v6.2.18-label-debug.2-0-ge57ea7951 go1.15.5

I just upgraded to teleport 8 but the issue still remains and the same output is given when trying to use the config file you provided (and the one in the related issue). 
As suggested, I tried using the configuration genereated by `tsh config` and in the end I had to change the command line to the following: 
`ssh -J vpn-strongswan.tele-test.dot.com admin@172.40.1.1`
It now works as expected.


When you state that it is expected behavior, isn't that something which should be documented or subject to a feature request ?
